add bucket authentication for server side encryption algorithm

nadavMiz opened the pull request:
### Explain the changes
expend bucket policy to enable statements containing conditions for authenticating based on server side encryption algorithm. providing base for farther adding other condition statements. enabling deny, accept bucket policies for stringEquals StringNotEquals and null conditions for server side encryptions

### Testing Instructions:
1. ceph s3 test: s3tests_boto3.functional.test_s3:test_bucket_policy_put_obj_s3_noenc

liranmauda commented on src/api/common_api.js:
```diff
@@ -252,6 +252,15 @@ module.exports = {
+        bucket_policy_condition: {
+            type: 'object',
+            properties: {
+                serverSideEncryption: {
```
Stile: 
this is a camelcase style, we are using snake_case
```suggestion
                server_side_encryption: {
```

aspandey commented on src/endpoint/s3/ops/s3_put_bucket_policy.js:
```diff
@@ -23,10 +23,38 @@ async function put_bucket_policy(req) {
+        parsed_condition[valid_conditions[condition]] = {};
+        for (const [condition_key, value] of Object.entries(obj)) {
+            if (!condition_key_map[condition_key]) {
+                console.error('put_bucket_policy: invalid condition key', condition_key);
```
Not only condition_key but also log "condition" which has this condition_key. This will have better information if it fails.

nadavMiz replied:
changed it

liranmauda commented on src/endpoint/s3/ops/s3_put_bucket_policy.js:
```diff
@@ -23,10 +23,38 @@ async function put_bucket_policy(req) {
 }
 
+function _parse_condition(conditions) {
+    const valid_conditions = {
```
Please add the link for the conditions in a comment:

```suggestion
    // https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html
    const valid_conditions = {
```

liranmauda replied:
We can change the name to reflact that those are the supported contions (currently it is only a subset of the list)

liranmauda commented on src/endpoint/s3/ops/s3_put_bucket_policy.js:
```diff
@@ -23,10 +23,38 @@ async function put_bucket_policy(req) {
+        'stringnotequals': 'string_not_equals',
+        'null': 'null'
+    };
+    const condition_key_map = {
```
Please add a link here.

liranmauda commented on src/endpoint/s3/ops/s3_put_bucket_policy.js:
```diff
@@ -23,10 +23,38 @@ async function put_bucket_policy(req) {
+    const parsed_condition = {};
+    for (const [condition, obj] of Object.entries(conditions)) {
+        if (!valid_conditions[condition]) {
+            console.error('put_bucket_policy: invalid condition', condition);
```
This Error is not indicating that we are not supporting the condition, for example if we get a condition that is valid such as `StringEqualsIgnoreCase` but we are not supporting it we should prompt the error that indicate that it is not supported.
also we can use dbg and not console
```suggestion
            dbg.error('put_bucket_policy: not supported condition', condition);
```

liranmauda commented on src/api/common_api.js:
```diff
@@ -252,6 +252,15 @@ module.exports = {
             }
         },
 
+        bucket_policy_condition: {
```
We should think if this is actually required. 
@dannyzaken WDYT?

jackyalbo reviewed (commented):
We should add unit tests to cover these changes in test_s3_bucket_policy. Please add.

nadavMiz commented:
> We should add unit tests to cover these changes in test_s3_bucket_policy. Please add.

I have added unit tests. also I have noticed I didn't handle returning the condition values to their original values in get _bucket_policy, so I have added that, and unit test for it

nadavMiz commented on src/endpoint/s3/ops/s3_put_bucket_policy.js:
```diff
@@ -23,10 +23,42 @@ async function put_bucket_policy(req) {
+    };
+    const parsed_condition = {};
+    for (const [condition, obj] of Object.entries(conditions)) {
+        if (!supported_conditions[condition]) {
```
the reason I added the map was to convert the s3 format to the DB format (same as all the other maps), since we convert the json to lowercase at the beginning of the function all the conditions seems to be a single word (the s3 style is camel case, so StringNotEqual will be converted to stringnotequal while I want to convert it to string_not_equal). so I don't think string manipulations will work in this case. I don't want to remove the conversion to lower case since it might hurt other functionality of the bucket policy. maybe I can just leave the conditions as single words in the DB as well.
In get-bucket-policy we already return all the policy values in lower case anyway

nadavMiz commented on src/endpoint/s3/ops/s3_put_bucket_policy.js:
```diff
@@ -23,10 +23,42 @@ async function put_bucket_policy(req) {
+function _parse_condition(conditions) {
+    // https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html
+    // TODO: this is a subset of the full list of conditions. should add more condtions in the future
+    const supported_conditions = {
```
its worth noting that since we convert the bucket policy to lower case. IgnoreCase wouldn't be different than regular stringEqual / stringNotEqual

jackyalbo replied:
so maybe we shouldn't just lowercase the whole bucket policy. we can move the lowercase to principle/action/reason and avoid doing it for conditions.

liranmauda reviewed (dismissed):
> We should think if this is actually required. 
> @dannyzaken WDYT?

I don't think that we should add bucket authentication for server-side encryption algorithm In a PR just for fixing a ceph test.

As @dannyzaken and I suggested before, this feels too big of a change as it is touching the authentication, and we should do it as a feature and not partially as part of a ceph test.

I think that we should close this PR. 

@nimrod-becker @dannyzaken @jackyalbo WDYT?

guymguym commented on src/api/common_api.js:
```diff
@@ -287,6 +321,32 @@ module.exports = {
+                                    string_not_like: {
+                                        $ref: '#/definitions/bucket_policy_condition'
+                                    },
+                                    null: {
```
we probably don't want to use a name such as `null` as a key - this will be error prone in the code.
Maybe call it null_condition or something like that?

nadavMiz replied:
Null is the key used by AWS. It should stay if we want to store the JSON as is. However, I can change it specifically if it is problematic

guymguym commented on src/api/common_api.js:
```diff
@@ -267,7 +301,7 @@ module.exports = {
                             },
                             effect: {
-                                enum: ['allow', 'deny'],
+                                enum: ['Allow', 'Deny'],
```
Is this change backward compatible?

guymguym commented on src/endpoint/s3/ops/s3_get_bucket_policy.js:
```diff
@@ -2,16 +2,38 @@
     return reply.policy;
 }
 
+function _parse_condition(conditions) {
```
is this needed? why can't we store the bucket policy json as is?

guymguym commented on src/endpoint/s3/ops/s3_put_bucket_policy.js:
```diff
@@ -10,21 +10,79 @@ const S3Error = require('../s3_errors').S3Error;
+    await req.object_sdk.put_bucket_policy({ name: req.params.bucket, policy: adapted_policy });
+}
+
+function _parse_condition(conditions) {
```
why are we parsing it? can't we store it as is?

guymguym commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -210,10 +210,14 @@ async function authorize_request_policy(req) {
+    const condition_elements = {};
+    condition_elements.encryption = s3_utils.parse_encryption(req);
+    condition_elements.bucket = req.params.bucket;
+    condition_elements.tagging = await req.object_sdk.get_object_tagging(req.params)?.tagging;
```
why do we always get the tagging here? this is always needed? can't we get it on demand?

nadavMiz committed:
change bucket policy schema to fit AWS json

Signed-off-by: nadav mizrahi <nadav.mizrahi16@gmail.com>

jackyalbo commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -0,0 +1,217 @@
+
+const predicate_map = {
+    'StringEquals': (request_value, policy_value) => request_value === policy_value,
+    'StringNotEquals': (request_val, policy_value) => request_val !== policy_value,
```
please use request_value instead of request_val in all the places in the map.
```suggestion
    'StringNotEquals': (request_value, policy_value) => request_value !== policy_value,
```

jackyalbo commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -0,0 +1,217 @@
+async function _is_server_side_encryption_fit(req, predicate, value) {
+    const encryption = s3_utils.parse_encryption(req);
+    const algorithm = encryption ? encryption.algorithm : null;
+    dbg.log1('bucket_policy: encrytpion fit?', value, algorithm, predicate(algorithm, value));
```
move calling the predicate outside of the function and print the result

jackyalbo commented on src/endpoint/s3/s3_bucket_policy_utils.js:
```diff
@@ -0,0 +1,217 @@
+    const reply = await req.object_sdk.get_object_tagging(req.params);
+    const tag = reply?.tagging?.find(element => (element.key === value.key));
+    const tag_value = tag ? tag.value : null;
+    dbg.log1('bucket_policy: object tag fit?', value, tag);
```
add here the predicate response like you did in the previous function (again without calling it twice)

jackyalbo commented on src/server/system_services/bucket_server.js:
```diff
@@ -513,6 +513,18 @@ function _validate_s3_policy(policy, bucket_name) {
+                    // parse to get only the condition key itself
+                    const key_parts = condition_key.split("/");
+                    if (!SUPPORTED_BUCKET_POLICY_CONDITIONS.includes(key_parts[0])) {
+                        throw new RpcError('MALFORMED_POLICY', 'Policy has invalid condition key', { detail: condition_key });
```
Should we return "malformed policy" in case of a condition we don't support? Maybe we should at least add something like this:
```suggestion
                        throw new RpcError('MALFORMED_POLICY', 'Policy has invalid or unsupported condition key', { detail: condition_key });
```
WDYT?
